upstream updates to chips-core from @barrystyle

satindergrewal opened the pull request:
These are all the updates fetched from @barrystyle's repo https://github.com/barrystyle/chips for upgrading chips-core to latest bitcoin-core 0.21+ (since master branch has more updates than version 0.21 branch) version.

The relevant github issue for these updates is here: https://github.com/chips-blockchain/chips/issues/10

I have tested these updates by compiling and it syncs to the network as expected.

- [ ] Task on taproot activation block height is yet to be sorted.

@who-biz @phm87 need your help reviewing this code, before merging to develop branch.

Commit:
port retarget based parameters across

Commit:
port genesis parameters, komodo components, cc and cryptoconditions

Commit:
emulate inconsistent segwit deployment

Commit:
reimplement getinfo with identical legacy output

Commit:
rebrand codebase as chips via automake

Commit:
correct some erroneous includes/linkage problems
correct some porting errors (serverchecker.cpp became sigcache.cpp)

Commit:
misc:

set default address type as per previous client
request more blocks, be less tolerant of slow peers
update icons for chips
tidy up splash screen

Commit:
continue porting komodo code adhoc

Commit:
update README.md

Commit:
port txMoMproof, calc_MoM and height_MoM function calls

Commit:
amend getinfo with the notarized parameters

Commit:
changes required from 0.21->master

Commit:
dont build the parts we arent using

Commit:
migrate getinfo to the new rpc definition style

Commit:
quick bitcoin.conf style generator

Commit:
update new static seeds

Commit:
doc: Remove unused section from release process

Commit:
remove cryptoconditions/cc and all linkages

Commit:
manually set taproot via block height

Commit:
setup temporary testnet

Commit:
actually accept the testnet chain

Commit:
correct issue with qt progress bar/headers

fanquake committed:
Merge bitcoin/bitcoin#22360: doc: Remove unused section from release process

fa051d3e5e50fb5726fc25464fef348255599e1e doc: Remove unused section from release process (MarcoFalke)

Pull request description:

  The section does not apply to us, because it is being done by the maintainers of the `bitcoin.org` website, see https://github.com/bitcoin-dot-org/Bitcoin.org/issues/3641#issuecomment-819586576 . Documentation about the process is in their repo: https://github.com/bitcoin-dot-org/bitcoin.org/blob/master/docs/adding-events-release-notes-and-alerts.md#release-notes

ACKs for top commit:
  jarolrod:
    ACK fa051d3e5e50fb5726fc25464fef348255599e1e
  Zero-1729:
    ACK fa051d3e5e50fb5726fc25464fef348255599e1e
  fanquake:
    ACK fa051d3e5e50fb5726fc25464fef348255599e1e

Tree-SHA512: eadc192958bb0d1009ad94c2ec79087b87514e3d19e2e6d759440392fa4abc432c512ba60692018e789c822a98a515a5f9a42966de4f8fe6b0a8b99470a76b5c

Commit:
correct erroneous include

sg777 commented:
To start with I'm doing functional testing on mainnet and testnet, will post my findings here.

who-biz commented:
The following was performed on chips testnet.  I am posting below for the purpose of documentation:

### Identified Issues

#### 1. `getinfo` command in `chips-cli` does not work, if a wallet hasn't been loaded

Note that we only receive the error message below on commands/methods that fetch wallet data.

```
chips-cli -testnet getinfo
error code: -18
error message:
No wallet is loaded. Load a wallet using loadwallet or create a new one with createwallet. (Note: A default wallet is no longer automatically created)
```
It might make things smoother for dPoW if we returned some information about the blockchain, in any case. If a wallet hasn't been loaded, we can just use placeholder values for balance, wallet version, etc. 

`iguana` checks `getinfo`, then `getblockchaininfo` if the first fails. By providing at least some information in absence of a wallet, we might gain some efficiency.  Particularly since iguana only requires this endpoint for `height, notarized_height` and other ntz-related data.

#### 2. When a wallet has been loaded, `getinfo` does not reflect balance properly

Comparing `getinfo` to `getwalletinfo`...

Output of `getinfo`:
```
chips-cli -testnet getinfo          
{
  "version": 219900,
  "protocolversion": 70016,
  "walletversion": 169900,
  "balance": 0.00000000,
  "blocks": 6175,
  "timeoffset": 0,
  "connections": 3,
  "proxy": "",
  "notarizedhash": "0000000000000000000000000000000000000000000000000000000000000000",
  "notarizedtxid": "0000000000000000000000000000000000000000000000000000000000000000",
  "notarized": 0,
  "prevMoMheight": 0,
  "notarized_MoMdepth": 0,
  "notarized_MoM": "0000000000000000000000000000000000000000000000000000000000000000",
  "difficulty": 0.01013083595512288,
  "testnet": true,
  "keypoololdest": 1626214667,
  "keypoolsize": 1999,
  "paytxfee": 0.00000000,
  "relayfee": 0.00001000,
  "errors": "This is a pre-release test build - use at your own risk - do not use for mining or merchant applications"
}
```

Output of `getwalletinfo`:

```
chips-cli -testnet getwalletinfo    
{
  "walletname": "test",
  "walletversion": 169900,
  "format": "bdb",
  "balance": 9199.99936000,
  "unconfirmed_balance": 0.00000000,
  "immature_balance": 100.00000000,
  "txcount": 187,
  "keypoololdest": 1626214667,
  "keypoolsize": 999,
  "hdseedid": "47f833b54ef50bc43bfbd938d119b10ccd18bf57",
  "keypoolsize_hd_internal": 1000,
  "paytxfee": 0.00000000,
  "private_keys_enabled": true,
  "avoid_reuse": false,
  "scanning": false,
  "descriptors": false
}
```

#### 3. Transactions are not being cleared from mempool when mining

SHossain found this issue, and I was able to reproduce.  From appearance of logs, transactions are not being included in `CreateNewBlock()`:

```
CreateNewBlock(): block weight: 860 txs: 0 fees: 0 sigops 400
```
Empty blocks being mined, despite mempool state:
```
chips-cli -testnet getrawmempool
[
  "0c74da539e6e61d00545d8b5634f88098006ba22d0951ccc18d06f5fd0f35aae",
  "f62da2a868405419d31c2e1bc1d91dfc206d1bc9bfc270d1e47b12bfba14ed87",
  "e12e05faeef0834e69b696fb47118a2714fad81045a5e443bb23a5967f15ad69",
  "3ef80d243597a53667c00bc428b3ed65cc9b107377500765206b89a77cddfbcc"
]
```

`getblocktemplate` also shows no transactions:

```
chips-cli -testnet getblocktemplate '{"rules": ["segwit"]}'
{
  "capabilities": [
    "proposal"
  ],
  "version": 536870912,
  "rules": [
    "csv",
    "taproot"
  ],
  "vbavailable": {
  },
  "vbrequired": 0,
  "previousblockhash": "0000004cc0a8af73a971c9f9fec013b5144541d657ddb7eb227c93bd6bdb464a",
  "transactions": [
  ],
  "coinbaseaux": {
  },
  "coinbasevalue": 5000000000,
  "longpollid": "0000004cc0a8af73a971c9f9fec013b5144541d657ddb7eb227c93bd6bdb464a876",
  "target": "00000062b5000000000000000000000000000000000000000000000000000000",
  "mintime": 1626301821,
  "mutable": [
    "time",
    "transactions",
    "prevblock"
  ],
  "noncerange": "00000000ffffffff",
  "sigoplimit": 20000,
  "sizelimit": 1000000,
  "curtime": 1626302182,
  "bits": "1d62b500",
  "height": 6245,
  "default_witness_commitment": "6a24aa21a9ede2f61c3f71d1defd3fa999dfa36953755c690689799962b48bebd836974e8cf9"
}
```

who-biz committed:
Fix broken balance display in getinfo

- Move 'getinfo' method into wallet section of cli

who-biz commented:
~~Please note that item 1 is no longer an issue, due to moving `getinfo` into wallet portion of RPC.~~

~~item 2 has been resolved and has a PR here: https://github.com/satindergrewal/bitcoin/pull/3.~~

~~Will begin working on mempool problems next.~~

who-biz committed:
Add notarization data to 'getblockchaininfo'

who-biz committed:
Always consider segwit txs when mining

- A less simple fix might be in order for this. Not sure if it matters for mainnet

who-biz committed:
Typo fix for Chips-Qt units

satindergrewal committed:
Merge pull request #6 from who-biz/qt-units-typo

Typo fix for Chips-Qt units

satindergrewal committed:
Merge pull request #5 from who-biz/blocktemplate-quickfix

Quick fix for `getblocktemplate`, gets mining and mempool working

satindergrewal committed:
Merge pull request #4 from who-biz/getbcinfo-ntzdata

Add notarization data to `getblockchaininfo`

satindergrewal committed:
Merge pull request #3 from who-biz/getinfo-fix

Fix broken balance display in `getinfo`

who-biz committed:
Fix broken calc_MoM RPC

who-biz committed:
Fix broken height_MoM method

who-biz committed:
Rebranding of help text, port change

satindergrewal committed:
Merge pull request #8 from who-biz/calcMoM

Fix broken calc_MoM command

satindergrewal committed:
Merge pull request #9 from who-biz/heightMoM

Fix broken height_MoM RPC

satindergrewal committed:
Merge pull request #10 from who-biz/help-text-branding

Rebranding of help text, port change

who-biz committed:
Fix fatal crash on 'getchaintips' call

who-biz committed:
Set an abritrary default txfee

- Removes need for user to manually call 'chips-cli settxfee x'

satindergrewal committed:
Merge pull request #13 from who-biz/getchaintips-fix

Fix fatal crash on 'getchaintips' call

satindergrewal committed:
Merge pull request #14 from who-biz/default-txfee

Set an abritrary default txfee

who-biz committed:
Fix 'pushKV ambiguous' error in compilation

- Error occuring because there is no conversion function for size_t to C string in pushKV variatns
- When conversion from size_t to uint64_t doesn't happen automatically on some systems, error is encountered
- Affects at least M1 macOS

satindergrewal committed:
Merge pull request #15 from who-biz/pushKV-fix-mac

Fix 'pushKV ambiguous' error in compilation

who-biz committed:
Add komodo_(dis)connect_block() function calls

who-biz committed:
Add definitions for ImportAddress, uncomment in kmd_validation015.h

- Things have changed quite a bit within scripts, particularly since the addition of Taproot
- Handling for CScriptIDs is somewhat different now too, so this should be checked for validity and proper function

satindergrewal committed:
Merge pull request #16 from who-biz/kmd-conn-disconn-block

Add komodo_(dis)connect_block() function calls

satindergrewal committed:
Merge pull request #17 from who-biz/importaddress

Add definitions for ImportAddress, uncomment in kmd_validation015.h

who-biz committed:
Set flat txfee of 10k sats

who-biz committed:
Use CHIPS legacy code for creating genesis block

- New upgrade added code that changed the manner in which genesis block was being constructed.  For ease of review, this commit patches in the old mechanics

satindergrewal committed:
Merge pull request #18 from who-biz/change-flatfee

Set flat txfee of 10k sats

satindergrewal committed:
Merge pull request #19 from who-biz/genesis-block

Use CHIPS legacy code for creating genesis block

Commit:
reuse the standard bitcoin genesis creation code with chips params

Commit:
match max op_return relay policy from legacy chips codebase

satindergrewal committed:
Merge pull request #21 from barrystyle/adapt-genesis

standardize genesis creation

satindergrewal committed:
Merge pull request #22 from barrystyle/opreturn-limits

increase op_return relay size

who-biz committed:
Fix validateaddress RPC, move into 'wallet' category

who-biz committed:
Use KMD-style addresses on testnet, for easier testing

satindergrewal committed:
Merge pull request #23 from who-biz/validateaddress-fix

Fix validateaddress RPC, move into 'wallet' category

satindergrewal committed:
Merge pull request #24 from who-biz/testnet-R-addrs

Use KMD-style addresses on testnet, for easier testing
